lib/phoenix_live_view/html_formatter.ex

```diff
@@ -552,7 +554,42 @@ defmodule Phoenix.LiveView.HTMLFormatter do
     [{:tag_block, name, attrs, block, %{mode: mode}} | list]
   end
 
-  defp may_set_preserve(buffer, _text), do: buffer
+  @non_ws_preserving_elements ["button"]
+
+  defp may_set_preserve_on_block(buffer, _text), do: buffer
+
+  defp may_set_preserve_on_text([{:text, text, meta}], :inline, tag_name)
+       when tag_name not in @non_ws_preserving_elements do
+    {mode, text} =
+      if meta.newlines == 0 and whitespace_around?(text) do
+        {:preserve, cleanup_extra_spaces(text)}
+      else
+        {:normal, text}
+      end
+
+    [{:text, text, Map.put(meta, :mode, mode)}]
+  end
+
+  defp may_set_preserve_on_text(buffer, _mode, _tag_name), do: buffer
+
+  defp whitespace_around?(text), do: :binary.first(text) in '\s\t' or :binary.last(text) in '\s\t'
+
+  defp cleanup_extra_spaces(text) do
+    text
+    |> cleanup_extra_spaces(0)
+    |> String.reverse()
+    |> cleanup_extra_spaces(0)
+    |> String.reverse()
+  end
+
+  defp cleanup_extra_spaces(<<char, rest::binary>>, count) when char in '\s\t',
+    do: cleanup_extra_spaces(rest, count + 1)
+
+  defp cleanup_extra_spaces(<<char, rest::binary>>, count) when count >= 1,
+    do: IO.iodata_to_binary([" ", char, rest])
+
+  defp cleanup_extra_spaces(<<char, rest::binary>>, _count),
+    do: IO.iodata_to_binary([char, rest])
```

Comment:
I think we could leverage `String` here:

```suggestion
  defp cleanup_extra_spaces_leading(text) do
    if :binary.first(text) in '\s\t' do
      IO.iodata_to_binary([" ", String.trim_leading(text)])
    else
      text
    end
  end

  defp cleanup_extra_spaces_trailing(text) do
    if :binary.last(text) in '\s\t' do
      IO.iodata_to_binary([String.trim_trailing(text), " "])
    else
      text
    end
  end
```